### DISCUSSION TITLE
Small file storage optimization（小文件存储优化）

### DISCUSSION THREAD

**Discussion by Dong-GuoSheng** [cubefs/cubefs#3584] [Q&A]
I wrote 5 million 1kb small files and found that the storage used 19G, but the size of 10 million small files is 4.768G. Each file smaller than 4kb actually uses 4kb for storage. Can I adjust the parameters to make the storage use the actual file size instead of using 4kb no matter how small the file is?
Because I have hundreds of millions of files every day, most of them are small files, so the storage expansion is quite serious.

我写入500万个1kb的小文件，发现存储使用了19G，但1000万个小文件是否大小为4.768G，每个小于4kb的文件都实际存储使用了4kb，能否通过调整参数让存储使用实际文件大小，不是无论文件多小都使用4kb？
因为每天我有上亿的文件，里面大部分都是小文件，这样存储膨胀的就比较厉害了。

**Comment by hooklee2000** [comment]
Use ext4 punch hole for tinyextent.  Ext4‘s default block size is 4096 byte, so default align to 4k. 
You need  format disks  as 1024 byte using mkfs.ext4 firstly. then modify the golang code fallocate() to align the offset and length. Be careful to test it.
 
The relationship between punching a hole in an ext4 file and the file system's block size is crucial because the hole punching operation works at the granularity of blocks. Here's how they are related:

1. Block Size in ext4
The block size in ext4 is determined when the file system is created (mkfs.ext4).
Common block sizes are 1 KiB, 2 KiB, 4 KiB, or 64 KiB, depending on system requirements and use cases.
You can check the block size of an existing file system using:
```bash
sudo blockdev --getbsz /dev/<device>
```
2. Hole Punching and Block Size
Block-Aligned Operations: Hole punching in ext4 requires the start offset (-o) and length (-l) to align with the file system's block boundaries. If they are not aligned, the operation may fail or only partially succeed.
Example: If the block size is 4 KiB and you specify an offset that is not a multiple of 4 KiB, the command will either throw an error or adjust to the nearest block boundary.
Granularity of Deallocation: Only full blocks can be deallocated when punching a hole. Partially filled blocks cannot be partially deallocated; they remain allocated until completely unused.
3. Practical Examples
Example 1: Block-Aligned Hole Punching
If the block size is 4 KiB, the offset and length must be multiples of 4 KiB:

```bash

fallocate -p -o $((1024 * 1024)) -l $((8 * 1024)) myfile
```
This punches a hole of 8 KiB starting at the 1 MiB offset. Both are multiples of 4 KiB.

Example 2: Non-Aligned Offset
If you attempt to punch a hole with a non-aligned offset or length:

```bash

fallocate -p -o 123 -l 4096 myfile
```
This operation will likely fail because 123 is not aligned to the block size (4 KiB).

Programmatic Example
In c, when using fallocate with FallocFLPunchHole , ensure the offset and length are block-aligned:

```cpp

#define BLOCK_SIZE 4096 // Example block size

off_t offset = 1024 * 1024; // Aligned to 4 KiB
off_t length = 8192;       // Multiple of 4 KiB
fallocate(fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE, offset, length);
```
4. Checking the Result
You can verify the file's actual disk usage with du after punching a hole:

```bash
du -h myfile
```
5. Implications of Block Size
Efficiency: Larger block sizes mean fewer metadata overheads but may lead to more unused space in sparse files if the hole is small or unaligned.
Alignment: Operations are simpler with larger block sizes since fewer offsets need to be considered.
Fragmentation: For workloads with frequent hole punching, a smaller block size can reduce wasted space but may increase fragmentation.
6. Advanced: Non-Default Block Sizes
You can create an ext4 file system with a non-default block size using mkfs.ext4:



mkfs.ext4 -b 4096 /dev/<device>
The -b option specifies the block size (e.g., 1 KiB, 2 KiB, or 4 KiB).

Summary
Block Alignment: Hole punching requires offset and length to be aligned with the file system block size.
Granularity: Only full blocks can be deallocated; partially filled blocks remain allocated.
Efficiency: Choose an appropriate block size for your file system based on the expected file and hole sizes.
Understanding and configuring the block size correctly can help optimize storage efficiency and performance in ext4 when using sparse files and hole punching.
